src/Tasks/LargeFileUploadTask.php

```diff
@@ -0,0 +1,345 @@
+        $baseUrl = rtrim($adapter->getBaseUrl(), '/');
+        $path = ltrim($url, '/');
+
+        $newUrl = "{$baseUrl}/$path";
+        $requestInformation->setUri($newUrl);
+        $requestInformation->httpMethod = HttpMethod::POST;
+        $requestInformation->setContentFromParsable($adapter, 'application/json', $uploadSessionBody);
+        return $adapter->sendAsync($requestInformation, [LargeFileTaskUploadSession::class, 'createFromDiscriminatorValue']);
+    }
+    /**
+     * @return RequestAdapter
+     */
+    public function getAdapter(): RequestAdapter {
+        return $this->adapter;
+    }
+
+    /**
+     * @return int
+     */
+    public function getChunks(): int {
+        return $this->chunks;
+    }
+
+    /**
+     * @return int
+     */
+    public function getUploadedChunks(): int {
+        return $this->uploadedChunks;
+    }
+
+    /**
+     * @param Parsable|null $uploadSession
+     * @throws Exception
+     */
+    public function uploadSessionExpired(?Parsable $uploadSession): bool {
+        $now = new DateTime((new DateTime('now'))->format(DateTimeInterface::ATOM));
+
+        $validatedValue = $this->checkValueExists($uploadSession ?? $this->uploadSession, 'getExpirationDateTime', ['ExpirationDateTime', 'expirationDateTime']);
+        if (!$validatedValue[0]) {
+            throw new Exception('The upload session does not contain an expiry datetime.');
+        }
+        $expiry = $validatedValue[1];
+
+        if ($expiry === null){
+            throw new InvalidArgumentException('The upload session does not contain a valid expiry date.');
+        }
+        $then = new DateTime($expiry->format(DateTimeInterface::ATOM));
+        $interval = $now->diff($then);
+
+        if ($interval->invert !== 0){
+            return true;
+        }
+        return false;
+    }
+    /**
+     * @throws Exception
+     */
+    public function upload(?callable $afterChunkUpload = null): void {
+
+        if ($this->uploadSessionExpired($this->uploadSession)){
+            throw new RuntimeException('The upload session is expired.');
+        }
+        $q = new SplQueue();
+
+        $start = 0;
+        $session = $this->nextChunk($this->stream, $start,max(0, min($this->maxChunkSize - 1,  $this->fileSize - 1)));
+        $q->enqueue($session);
+
+        while(!$q->isEmpty()){
+            /** @var Promise $front */
+            $front = $q->dequeue();
+
+            $front->then(function (LargeFileTaskUploadSession $session) use (&$q, $afterChunkUpload){
+                $nextRange = $session->getNextExpectedRanges();
+                $this->uploaded = (int)explode('-', $nextRange[0] ?? ($this->fileSize.'-'))[0];
+                if (empty($nextRange)) {
+                    echo "Upload finished!!!!\n";
+                    return $session;
+                }
+                $this->uploadedChunks++;
+                if (!is_null($afterChunkUpload)) {
+                    $afterChunkUpload($this);
```

Comment:
Expected parameters for the callback: `rangeOfFileUploaded, additionalParamters` - https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/tasks/FileUploadTask.md

Reply:
@SilasKenneth  was referring to the `progress` callback signature in this section...

![image](https://user-images.githubusercontent.com/10958912/212840976-7dbc3cba-1d5f-4977-8f55-89153078c3bd.png)

Thoughts?

Reply:
Hi @Ndiritu ,
Updated this.